[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29104 [bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256**

We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.


Single Core:
(Note that our benchmark generates batch_size=47 for first case and batch_size=56 for the second case. In spite of that, the vectorized version is still faster than the original reference C version without vectorization.)
- Before the PR:
```
native_layer_norm        0.81%            5.884ms          0.81%            5.884ms          122.580us        NaN              0.000us          0.000us          48               [[47, 1, 1024], [1024], [1024]]
```

- After the PR:
```
native_layer_norm        0.68%            5.053ms          0.68%            5.053ms          105.272us        NaN              0.000us          0.000us          48               [[56, 1, 1024], [1024], [1024]]
```


20 Cores:
- Before the PR:
```
native_layer_norm        1.65%            41.682ms         1.65%            41.682ms         868.365us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```


- After the PR:
```
native_layer_norm        1.34%            33.829ms         1.34%            33.829ms         704.771us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

jianyuh committed:
[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256

We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

BIT-silence commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -32,15 +36,28 @@ void LayerNormKernelImplInternal(
   T* mean_data = mean->data_ptr<T>();
   T* rstd_data = rstd->data_ptr<T>();
   const T c = T(1) / static_cast<T>(N);
+  constexpr int64_t VEC_SIZE = vec256::Vec256<T>::size();
```
It will be better to name it as kVecSize.

jianyuh replied:
Updated. Thanks!

BIT-silence commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -32,15 +36,28 @@ void LayerNormKernelImplInternal(
-      for (int64_t j = 0; j < N; ++j) {
+      vec256::Vec256<T> mean_vec(0);
+      vec256::Vec256<T> rstd_vec(0);
+      int64_t j = 0;
```
Put j declaration inside for-loop?

jianyuh replied:
The declaration outside loop can keep the value of j in the next loop (See line 60).

BIT-silence commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -109,15 +126,32 @@ void LayerNormBackwardKernelImplInternal(
+        const vec256::Vec256<T> x_vec = vec256::Vec256<T>::loadu(&X_ptr[j]);
+        const vec256::Vec256<T> gamma_vec = gamma_null
+            ? vec256::Vec256<T>(1)
+            : vec256::Vec256<T>::loadu(&gamma_data[j]);
```
```suggestion
            : vec256::Vec256<T>::loadu(gamma_data + j);
```

jianyuh replied:
Updated. Thanks!

BIT-silence commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -109,15 +126,32 @@ void LayerNormBackwardKernelImplInternal(
-      for (int64_t j = 0; j < N; ++j) {
+      vec256::Vec256<T> ds_vec(0);
+      vec256::Vec256<T> db_vec(0);
+      int64_t j = 0;
```
ditto.

jianyuh replied:
The declaration outside loop can keep the value of j in the next loop (See line 168).

BIT-silence replied:
Actually there is no gain at all in modern compiler but lead to worse readability.

jianyuh replied:
Agreed: there is little gain on the modern compiler. In terms of readability, I feel it should have better readability as it avoids the duplication and carries the value of j to the next loop. Check https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/functional.h#L30-L36 for a similar code piece (`d` is defined before the loop and the value of `d` is carried to the later `if` statement).

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/cpu/vec256/vec256_double.h:
```diff
@@ -307,11 +307,16 @@ inline void convert(const double* src, double* dst, int64_t n) {
+template <>
+Vec256<double> inline fmadd(const Vec256<double>& a, const Vec256<double>& b, const Vec256<double>& c) {
+  return a * b + c;
+}
```
I don't think this is a good idea, it's changing numerics based on the available vector extension, and that is not clear to a user of vec256

Can we just do this at the callsites?

jianyuh replied:
Sure: separate that part into a new PR: https://github.com/pytorch/pytorch/pull/29154

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jianyuh commented:
The test failure is weird: I can reproduce it on a Skylake machine before my PR with
```
python run_test.py -i nn -- TestNN.test_LayerNorm_1d_no_elementwise_affine_eval
```

Output:
```
$ python run_test.py -i nn -- TestNN.test_LayerNorm_1d_no_elementwise_affine_eval
which: no nvcc in (/root/miniconda/bin:/root/miniconda/condabin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/facebook/ops/scripts:/usr/facebook/scripts:/opt/local/bin:/usr/facebook/scripts:/usr/facebook/scripts/db:/root/bin)
Test executor: ['/root/miniconda/bin/python']
Running test_nn ... [2019-11-09 21:25:55.865797]
F
======================================================================
FAIL: test_LayerNorm_1d_no_elementwise_affine_eval (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 7344, in <lambda>
    add(test_name, lambda self, test=test: test(self))
  File "/root/jhuang_test/pytorch/test/common_nn.py", line 3511, in __call__
    self.test_noncontig(test_case, module, input)
  File "/root/jhuang_test/pytorch/test/common_nn.py", line 3570, in test_noncontig
    test_case.assertEqual(out, output)
  File "/root/jhuang_test/pytorch/test/common_utils.py", line 736, in assertEqual
    assertTensorsEqual(x, y)
  File "/root/jhuang_test/pytorch/test/common_utils.py", line 706, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(76.1324, grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :

----------------------------------------------------------------------
Ran 1 test in 0.009s

FAILED (failures=1)
Traceback (most recent call last):
  File "run_test.py", line 455, in <module>
    main()
  File "run_test.py", line 447, in main
    raise RuntimeError(message)
RuntimeError: test_nn failed!
```
On a Broadwell machine, for the code version both before and after my PR, there is no such failure.

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jianyuh commented:
Fixed the issue by reusing `vec256::reduce_all` and `vec256::map_reduce_all`.

jamesr66a reviewed (approved):
LGTM. Does this improve some benchmarks?

jianyuh commented:
> LGTM. Does this improve some benchmarks?

Will check the performance before landing. Thanks!

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]

jianyuh commented:
Update the performance number in the summary.

jianyuh committed:
Update on "[bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256"


We would like to provide the vectorized implementation for layer norm. This PR reuses https://github.com/pytorch/pytorch/pull/23349.


Single Core:
(Note that our benchmark generates batch_size=47 for first case and batch_size=56 for the second case. In spite of that, the vectorized version is still faster than the original reference C version without vectorization.)
- Before the PR:
```
native_layer_norm        0.81%            5.884ms          0.81%            5.884ms          122.580us        NaN              0.000us          0.000us          48               [[47, 1, 1024], [1024], [1024]]
```

- After the PR:
```
native_layer_norm        0.68%            5.053ms          0.68%            5.053ms          105.272us        NaN              0.000us          0.000us          48               [[56, 1, 1024], [1024], [1024]]
```


20 Cores:
- Before the PR:
```
native_layer_norm        1.65%            41.682ms         1.65%            41.682ms         868.365us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```


- After the PR:
```
native_layer_norm        1.34%            33.829ms         1.34%            33.829ms         704.771us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```

Differential Revision: [D18293522](https://our.internmc.facebook.com/intern/diff/D18293522/)

[ghstack-poisoned]
